src/intel/sse.rs

```diff
@@ -0,0 +1,1400 @@
+/// let c = unpack_low_m128(a, b).to_array();
+/// assert_eq!(c, [1.0, 5.0, 2.0, 6.0]);
+/// ```
+#[must_use]
+#[inline(always)]
+pub fn unpack_low_m128(a: m128, b: m128) -> m128 {
+  m128(unsafe { _mm_unpacklo_ps(a.0, b.0) })
+}
+
+/// Bitwise `a ^ b`.
+/// ```
+/// # use safe_arch::*;
+/// let a = m128::from_array([1.0, 0.0, 1.0, 0.0]);
+/// let b = m128::from_array([1.0, 1.0, 0.0, 0.0]);
+/// let c = xor_m128(a, b).to_array();
+/// assert_eq!(c, [0.0, 1.0, 1.0, 0.0]);
+/// ```
+#[must_use]
+#[inline(always)]
+pub fn xor_m128(a: m128, b: m128) -> m128 {
+  m128(unsafe { _mm_xor_ps(a.0, b.0) })
+}
+
+//
+// Here we define the Operator Overloads for `m128`. Each one just calls the
+// correct function from above. By putting the impls here and not with the
+// `m128` type we theoretically would be able to build the crate safely even if
+// there's no `sse` feature enabled. You'd just have a `m128` type without the
+// operator overloads is all. Not that the standard Rust distribution can build
+// properly without `sse` enabled, but maybe you're using a custom target or
+// something. It doesn't really put us out of our way, so it doesn't hurt to try
+// and accommodate the potential use case.
+//
+
+impl Add for m128 {
+  type Output = Self;
+  fn add(self, rhs: Self) -> Self {
+    add_m128(self, rhs)
+  }
+}
+impl AddAssign for m128 {
+  fn add_assign(&mut self, rhs: Self) {
+    *self = *self + rhs;
+  }
+}
+
+impl BitAnd for m128 {
+  type Output = Self;
+  fn bitand(self, rhs: Self) -> Self {
+    and_m128(self, rhs)
+  }
+}
+impl BitAndAssign for m128 {
+  fn bitand_assign(&mut self, rhs: Self) {
+    *self = *self & rhs;
+  }
+}
+
+impl BitOr for m128 {
+  type Output = Self;
+  fn bitor(self, rhs: Self) -> Self {
+    or_m128(self, rhs)
+  }
+}
+impl BitOrAssign for m128 {
+  fn bitor_assign(&mut self, rhs: Self) {
+    *self = *self | rhs;
+  }
+}
+
+impl BitXor for m128 {
+  type Output = Self;
+  fn bitxor(self, rhs: Self) -> Self {
+    xor_m128(self, rhs)
+  }
+}
+impl BitXorAssign for m128 {
+  fn bitxor_assign(&mut self, rhs: Self) {
+    *self = *self ^ rhs;
+  }
+}
+
+impl Div for m128 {
+  type Output = Self;
+  fn div(self, rhs: Self) -> Self {
+    div_m128(self, rhs)
+  }
+}
+impl DivAssign for m128 {
+  fn div_assign(&mut self, rhs: Self) {
+    *self = *self / rhs;
+  }
+}
+
+impl Mul for m128 {
+  type Output = Self;
+  fn mul(self, rhs: Self) -> Self {
+    mul_m128(self, rhs)
+  }
+}
+impl MulAssign for m128 {
+  fn mul_assign(&mut self, rhs: Self) {
+    *self = *self * rhs;
+  }
+}
+
+impl Neg for m128 {
+  type Output = Self;
+  fn neg(self) -> Self {
+    sub_m128(zeroed_m128(), self)
+  }
+}
+
+impl Not for m128 {
+  type Output = Self;
+  /// Not a direct intrinsic, but it's useful and the implementation is simple
+  /// enough.
+  ///
+  /// This performs an `xor` with an all-1s bit pattern.
+  fn not(self) -> Self {
+    let all_bits = splat_m128(f32::from_bits(u32::MAX));
+    self ^ all_bits
```

Comment:
well we want to be nan safe, but is a zero and a cmp cheaper than splatting a const? we'll have to godbolt it probably

Reply:
Yeah, Godbolt to be sure. Creating zero is 1 cycle, comparing it is 1, both operations are open for ILP. So that's 2 cycles to create a `-1` vector out of thin air. If LLVM doesn't figure out a similar trick, the current code will be a memory load, which moves over from hot I-cache to potentially cold D-cache. An L1 cache hit takes 2 cycles. And an L3 cache miss… well…